### ISSUE TITLE
Missing Java classes with proguard

### ISSUE THREAD

**Issue by toxikman** [google/play-unity-plugins#37]
When creating a ReviewManager in Unity 2018.4.24f1, I get the following exception:

_AndroidJavaException: java.lang.ClassNotFoundException: com.google.android.play.core.review.ReviewManagerFactory_

This application is built with gradle and proguard enabled. When I manually copy the proguard rules from the packages "Proguard" folder to my proguard-user.txt file, the exception goes away. The proguard rules in the "Google Play In-App Reviews" UPM package and other Google Play UPM package dependencies are not referenced in the generated build.gradle script. The documentation doesn't say anything about having to manually adding these to our proguard user rules file.  Unity's [package layout documentation](https://docs.unity3d.com/Manual/cus-layout.html) says nothing about "Proguard" folders being a standard way to include proguard rules for Android builds, nor does the EDM4U system seem to allow for this either.

**Comment by thanwah** [comment]
Thanks @toxikman for filing this issue. It sounds like you were able to get the proguard config working as intended, but it's a good point that we don't have this process documented clearly. I've created an issue on our internal tracker to improve the documentation.

**Comment by toxikman** [comment]
Thanks @thanwah . That's good to know that the proguard rules have to be manually added. Currently I'm just copy-pasting the text from the package Proguard folder, however, there are some problems with this:

1. If you select just the In-App Review package from the Unity Package Manager UI, this will pull in 3 other packages due to dependencies (common, core, and appbundle). Because these are "dependencies," Unity hides these packages in the Project browser. But the com.google.play.core package also contains Proguard dependencies. Developers have to know this and then find the Unity PackageCache in order to copy these hidden proguard rules.
1. I love that you guys made these SDKs into Unity packages, but the advantage of packages is that you can update them easily. If the proguard rules change in the new versions, this could pose a problem if we forget to replace the rules in our proguard-user.txt file. If we could reference the packages' proguard rules files directly from the gradle template, that would mean we always get the correct proguard rules whenever we update and don't have to manually copy anything. Could this be done?

**Comment by chrisyarbrough** [comment]
Is there any updated documentation for the process of setting up the Proguard files for the Play plugins within Unity? I also managed to get it working, but it was really a struggle going through all of the pain before I landed on the solution.

The way I did it was:
- In ProjectSettings > Publishing Settings > Minify > Enable Release (Proguard)
- Enable Custom Proguard File
- Manually copy the contents of multiple plugin's proguard files into the user file (e.g. com.google.play.assetdelivery/asset_delivery.txt)

So I wonder, is this the intented workflow? Is there no way for the plugins to automate this? Or would Unity have to change things to make this possible?

**Comment by jamcohen** [comment]
Hi @chrisyarbrough,

Yes, that's still the intended workflow. It's not ideal, but Unity doesn't provide a way to specify multiple proguard.txt files. We considered having the plugin merge the individual proguard files into a single version, but decided against it because a couple edge cases added complexity:
1. The system would have to handle adding/updating and removing plugins.
2. The system would have to account for a user's existing proguard.txt file.

We appreciate you calling this out though. We'll explore other ways of making this easier.

**Comment by toxikman** [comment]
> Yes, that's still the intended workflow. It's not ideal, but Unity doesn't provide a way to specify multiple proguard.txt files. We considered having the plugin merge the individual proguard files into a single version, but decided against it because a couple edge cases added complexity:

Could you have package-specific proguard rules using the existing AAR proguard.txt support? By creating an unique AAR file with proguard rules that list the classes that are referenced by the Unity plug-in scripts. I believe Unity or gradle merges these rule files from AARs at build-time.

**Comment by chrisyarbrough** [comment]
@jamcohen Thanks for letting me know! I think it's ok doing this by hand as a user, but it would be great if this was documented better. It might be obvious for Android veterans, but I believe many Unity users will have troubles realizing that minification causes stripping issues, so I think this info should be part of a getting started guide.

Or, maybe it would be possible for the plugins which know that they need custom proguard files, to show a Unity editor dialog box the first time an enabled minify setting is detected. It could be one of those "don't show again" boxes that remind users to setup their files, potentially also remind them of cleaning them once plugins were removed etc.

**Comment by jamcohen** [comment]
@toxikman, are you suggesting we include a mostly empty AAR in each package that only includes a proguard.txt file? I hadn't considered that before. I didn't think the proguard rules would apply to classes outside of the AAR but it's worth trying. Thanks for the suggestion.

@chrisyarbrough, both good points. If we can't make this automatic, a dialog is a good compromise.

**Comment by toxikman** [comment]
> @toxikman, are you suggesting we include a mostly empty AAR in each package that only includes a proguard.txt file? I hadn't considered that before. I didn't think the proguard rules would apply to classes outside of the AAR but it's worth trying. Thanks for the suggestion.

Correct. I'm pretty sure this works because my game uses a lot of ad networks that include proguard.txt files in their AAR files because the main class entry-points are typically referenced by unity scripts using AndroidJavaObject. If proguard says to keep the class, then the class is kept, regardless of whether it's referenced from within the AAR or anywhere else.

> @chrisyarbrough, both good points. If we can't make this automatic, a dialog is a good compromise.
Google Play Game Services plugin has the same issues. Maybe you guys could agree on a common Tools menu or UI where proguard rules could be added. Like a checkbox "Add proguard rules to Assets/Plugins/Android/proguard-user.txt"
